lib/std/io_uring.zig

```diff
@@ -0,0 +1,841 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2015-2020 Zig Contributors
+// This file is part of [zig](https://ziglang.org/), which is MIT licensed.
+// The MIT license requires this copyright notice to be included in all copies
+// and substantial portions of the software.
+const builtin = @import("builtin");
+const std = @import("std");
+const assert = std.debug.assert;
+const os = std.os;
+const linux = os.linux;
+const mem = std.mem;
+const net = std.net;
+const testing = std.testing;
+
+pub const io_uring_params = linux.io_uring_params;
+pub const io_uring_cqe = linux.io_uring_cqe;
+
+// TODO Update linux.zig's definition of linux.io_uring_sqe:
+// linux.io_uring_sqe uses numbered unions, i.e. `union1` etc. that are not future-proof and need to
+// be re-numbered whenever new unions are interposed by the kernel. Furthermore, Zig's unions do not
+// support assignment by any union member directly as in C, without going through the union, so the
+// kernel adding new unions would also break existing Zig code.
+// We therefore use a flat struct without unions to avoid these two issues.
+// Pending https://github.com/ziglang/zig/issues/6349.
+pub const io_uring_sqe = extern struct {
+    opcode: linux.IORING_OP,
+    flags: u8 = 0,
+    ioprio: u16 = 0,
+    fd: i32 = 0,
+    off: u64 = 0,
+    addr: u64 = 0,
+    len: u32 = 0,
+    opflags: u32 = 0,
+    user_data: u64 = 0,
+    buffer: u16 = 0,
+    personality: u16 = 0,
+    splice_fd_in: i32 = 0,
+    options: [2]u64 = [2]u64{ 0, 0 }
+};
+
+// TODO Add to zig/std/os/bits/linux.zig:
+const IORING_SQ_CQ_OVERFLOW = 1 << 1;
+
+comptime {
```

Comment:
Is this a unit test?

Reply:
No, these are fail-fast safety checks in case something goes haywire with the struct layouts (compiler bug, something changes in linux.zig etc.). This should never happen but it's critical for the mmap calls and ring pointers from a security point of view. We need those offsets to be exactly that, and we need e.g. the sqe struct to be exactly 64 bytes. I know not everyone likes assertions compared to unit tests, but I think it's good to have these here.

Reply:
Its okay to have these as is. A test would also be okay.
```zig
test "assert struct sizes" {
    comptime {
        // assert
    }
}

Reply:
Actually users would be reaching into this file so these asserts should run on condition `if (supported) {...}`

Reply:
The asserts will always pass on any version of Linux, since it's only checking structs and constants. It's not dependent on any kernel version.

Reply:
Are you sure? Are there any fields that may change size on different architectures? Do these constants also architecture independent?